LFS-641: Introduce Close button to all forms; LFS-722: Fields at the bottom of the form are inaccessible

marta- opened the pull request:
[LFS-641](https://phenotips.atlassian.net/browse/LFS-641): Introduce Close button to all forms
[LFS-722](https://phenotips.atlassian.net/browse/LFS-722): Fields at the bottom of the form are inaccessible

**This PR is back in draft mode due to leftover changes that were already merged, and because of known issues with pagination.**

LFS635 (Autosave form data when the user navigates away from a form page) was merged separately.
LFS272 (Forms are periodically saved automatically as the user fills out information) is currently excluded due to ongoing discussions about the right way to trigger saving.

marta- committed:
LFS-641: Introduce FINISH button to all forms

marta- committed:
LFS-272: Forms are periodically saved automatically as the user fills out information
First iteration, no UI feedback

sashaandjic reviewed (approved):
tested
all good ( auto save and FINISH bttn)

marta- committed:
[cleanup] Removed unused state

marta- committed:
LFS-272: Forms are periodically saved automatically as the user fills out information
* Added "Saving..." "Saved timestamp" in the UI
* Ensured autosave is always triggered when navigating away within the application

marta- committed:
LFS-635: Autosave form data when the user navigates away from a form page

marta- commented:
After interacting with the forms for many hours, I am now convinced "Done" is a better than "Finish" for (LFS-641). Thoughts? @sashaandjic @k-pace @ccmbioinfo/lfs-devs

sashaandjic commented:
retested; all good

as for the DONE/FINISH bttn question, now I'm thinking:
- no button at all (when there is no more pages disable NEXT bttn) OR
- CLOSE or red big (X) bttn + we should have it elsewhere too (charts, variants, admin forms)

k-pace commented:
I like 'Close' if possible over done - if the Next button can become 'Close' (or even Save and Close, just to make it clear to the user that nothing will be lost, even though auto-saving is always happening, sometimes users are nervous folk) when there are no more pages.
Re: back button, I actually like that it's visually differentiated from the next - it'll be a feature used less often so attention should be drawn to next rather than going backward I would think

marta- committed:
LFS-272: Forms are periodically saved automatically as the user fills out information
 Avoid some redundant autosave calls

sashaandjic commented:
The advantage of a separate close button  (big red  X) is that user does not need to go to the last page in order to close the form. Sometimes they may want to quickly see something on the first few pages and go back to where they were before.
I don't like clicking outside of the current form/chart as it will not take me back where I was initially. Many times I wished there was a BACK button, but CLOSE could do it as well.

k-pace commented:
Maybe the right solution is that we place a persistent 'Close' button beside 'Next' so that the user always has the option to get out of where they are?

sashaandjic commented:
Actually I would place it right next to the form title. 
![image](https://user-images.githubusercontent.com/23243244/99819410-714f4b00-2b1d-11eb-8e6b-0b870a097600.png)

marta- commented:
I like the idea of separating next and close.

Red x makes me think of a dangerous action. A gray x like what we have on dialogs may be more appropriate if we go this way.

sashaandjic commented:
re-tested

marta- committed:
LFS-272: Forms are periodically saved automatically as the user fills out information

Data is now saved with every single change made to the form instead of every 10s.

marta- commented:
@sashaandjic The latest commit eliminates the double autosave issue. Data is now saved with every single change of the form instead of every 10s. (note to self: we can probably even remove the `window.onbeforeunload` listener now). Things to test thoroughly:
* that data gets saved with every change
* what happens when session expires and user gets logged out
* what happens when internet connection is slow (use browser dev tools to throttle connection)
* what happens when user goes offline or server goes down (is feedback to the user clear/useful)

sashaandjic reviewed (approved):
tested

marta- committed:
LFS-272: Forms are periodically saved automatically as the user fills out information
Reorganize effects

marta- committed:
LFS-641: Introduce Close button to all forms
LFS-722: Fields at the bottom of the form are inaccessible

* Replaced Save/Finish button with (x) for Close
* Separated the pagination Next button from the final Save/Close
* Moved final save functionality from FormPagination back to Form

sdumitriu commented:
This is terribly slow on my (fast) computer. Typing is very sluggish, letters show up a couple of seconds after I typed them.

k-pace commented:
- I really like the new flow, especially the progress bar beside 'Next'

marta- committed:
Removing periodic autosave (LFS-272) which in its last version caused the UI to be extremely sluggish in Firefox.

marta- committed:
[misc] Fixed merge arifact, made some changes easyer to read

sashaandjic reviewed (approved):
tested

marta- committed:
LFS-641: Introduce Close button to all forms
Changed the icon for "save and close" from x to checkmark.

sashaandjic commented:
re-tested

marta- committed:
Merge branch 'dev' into LFS-641

marta- commented:
LFS-722 (Fields at the bottom of the form are inaccessible) was fixed separately in #398.

This PR no longer works.

LFS-641 (Introduce Close button to all forms) and LFS-272 (Forms are periodically saved automatically as the user fills out information) will have to be fixed at the same time for a decent user experience.
